### DISCUSSION TITLE
Using a date with multiple values when defining population

### DISCUSSION THREAD

**Discussion by elsie-h** [opensafely/documentation#408] [Ideas]
I'm working on a project in which I want to select individuals who have not been vaccinated in the 12 weeks following their eligibility date (project not yet approved so in my personal GitHub repo - can give access on request) . Therefore, when selecting the population I want `patients.registered_as_of(elig_date)`, and let's assume there are 10 different possible values for `elig_date`. I'll also want to use `elig_date` when defining other variables in the study definition.

I've allowed for this by creating a different study definition for each `elig_group`, which was done by adapting @wjchulme's [create-project.R](https://github.com/opensafely/comparative-ve-research/blob/main/create-project.R) such that the study definitions are created from a file called `template_study_definition.py`:

```
# define number of eligibility groups
n_elig_groups <- 10

create_study_definitions<- 
  str_c("#!/bin/sh",
        "# Run this script to create a study_definition for each eligibility date",
        "",
        str_c("for i in {1..",n_elig_groups,"}; do"),
        "sed -e \"s;%placeholder_elig_date%;elig_date_$i;g\" -e \"s;%placeholder_elig_grp%;$i;g\" ./analysis/template_study_definition.py > ./analysis/study_definition_$i.py;",
        "done;", sep = "\n")

create_study_definitions %>%
  writeLines(here::here("analysis/create_study_definitions.sh"))

# create study definitions from template_study_definition.py
check_create <- try(processx::run(command="bash", args= "analysis/create_study_definitions.sh"))

if (class(check_create)=="try-error") stop("Study definitions not created.")
``` 

Then I can create the actions for the project.yaml using the `action` function defined in [create-project.R](https://github.com/opensafely/comparative-ve-research/blob/main/create-project.R).

```
stud_def_actions <-  lapply(1:n_elig_groups,
                            function(x)
                              action(
                                name = str_c("elig_grp_", x),
                                run = str_c("cohortextractor:latest generate_cohort --study-definition study_definition_", x),
                                needs = list("design"),
                                highly_sensitive = list(
                                  cohort = str_c("output/input_",x,".csv")
                                )
                              ))
```

I was chatting to @wjchulme about this in case he knew of a more elegant way to do this, and he recommended posting it on here to see if anyone else has as a better solution/can use my solution if it's useful!

(I appreciate this might be a bit difficult to follow without access to the repo so please let me know and I can give you access!)

**Comment by wjchulme** [comment]
Worth pointing out that one failed attempt at solving this neatly was to use `patients.categorised_as` to create the group-specific date variable:

```
population = patients.registered_as_of("elig_date"),
jcvi_group = patients.categorised_as(
   {
      "group 1": " age>=80 OR carehomeresident OR carehomestaff",
      "group 2": "healthcareworker",
      "group 3": "age>=75",
      ...
   }
),
elig_date = patients.categorised_as(
   {
      "2020-12-08": "jcvi_group = 'group 1'",
      "2020-12-08": "jcvi_group = 'group 2'",
      "2021-01-04": "jcvi_group = 'group 3'",
      ...
   }
),
```
It doesn't work because `patient.categorised_as` returns categoricals, which won't be implicitly interpreted as dates.

**Reply by evansd** [reply]
@wjchulme's proposal looks to me like The Right Way of solving the problem except, as you point out, it doesn't work for very annoying typing reasons. This is, yet again, one of those issues that cohortextractor-v2 was designed to avoid but that doesn't help you much in the here and now.

However, it might be possible to tweak the `categorised_as` function a bit to support returning different types. Off the top of my head I can't decide whether that would be really quite straightforward or hideously difficult so I'll have to look into it more. But it feels worth at least some initial investigation as if it is possible it might provide an elegant solution to quite a few problems like this.

**Reply by elsie-h** [reply]
Thanks for looking into this, @evansd. As I'm new to working with OpenSAFELY I've not run a study definition yet on the TPP data, but I'm concerned that the workaround that I posted will be painfully slow as I currently have 22 eligibility groups.

**Reply by evansd** [reply]
I've fixed the problem with using dates in `categorised_as()` so something like @wjchulme's proposed solution should now work:
https://github.com/opensafely-core/cohort-extractor/pull/656

One minor thing in passing though: you'll need to take the spaces out of the category names and use e.g. `group_1` rather than `group 1` or it will complain when you try to use them in an expression.
